integrations/git_annex_test.go

```diff
@@ -0,0 +1,940 @@
+// Copyright 2022 The Gitea Authors. All rights reserved.
+// Use of this source code is governed by a MIT-style
+// license that can be found in the LICENSE file.
+
+// I should test both:
+// - symlink annexing
+// - smudge annexing
+// - http annexing
+// - ssh annexing
+//
+// and then cross all that with testing different combinations of permissions
+// ..yeah? Is that a reasonable thing to do?
+
+// it would also be good, probably, to test how push-to-create interacts with git-annex
+
+package integrations
+
+import (
+	"github.com/stretchr/testify/require"
+	"testing"
+
+	"bytes"
+	"errors"
+	"fmt"
+	"io"
+	"io/fs"
+	"math/rand"
+	"net/url"
+	"os"
+	"path"
+	"path/filepath"
+	"regexp"
+	"strings"
+
+	"code.gitea.io/gitea/models/perm"
+	repo_model "code.gitea.io/gitea/models/repo"
+	"code.gitea.io/gitea/modules/git"
+	"code.gitea.io/gitea/modules/setting"
+	api "code.gitea.io/gitea/modules/structs"
+	"code.gitea.io/gitea/modules/util"
+
+	"time" // DEBUG
+)
+
+func TestGitAnnex(t *testing.T) {
+	/*
+		// TODO: look into how LFS did this
+		if !setting.Annex.Enabled {
+			t.Skip()
+		}
+	*/
+
+	trueBool := true // this is silly but there's places it's needed
+	falseBool := !trueBool
+
+	// Some guidelines:
+	// a APITestContext is an awkward union of session credential + username + target repo
+	// which is assumed to be owned by that username; if you want to target a different
+	// repo, you need to edit its .Reponame or just ignore it and write "username/reponame.git"
+
+	onGiteaRun(t, func(t *testing.T, u *url.URL) {
+		defer annexUnlockdown() // workaround https://git-annex.branchable.com/internals/lockdown/
```

Comment:
Ah, no, I don't. I thought I did but I was confused. If I drop this, and add a print to RemoveAll, you can see the test harness uses `util.RemoveAll()` to clean up previous tests before it starts:

<details><summary>patch</summary>

```patch
diff --git a/integrations/git_annex_test.go b/integrations/git_annex_test.go
index 313cb2875..40983ca44 100644
--- a/integrations/git_annex_test.go
+++ b/integrations/git_annex_test.go
@@ -23,12 +23,10 @@ import (
        "errors"
        "fmt"
        "io"
-       "io/fs"
        "math/rand"
        "net/url"
        "os"
        "path"
-       "path/filepath"
        "regexp"
        "strings"
 
@@ -59,7 +57,6 @@ func TestGitAnnex(t *testing.T) {
        // repo, you need to edit its .Reponame or just ignore it and write "username/reponame.git"
 
        onGiteaRun(t, func(t *testing.T, u *url.URL) {
-               defer annexUnlockdown() // workaround https://git-annex.branchable.com/internals/lockdown/
 
                t.Run("Public", func(t *testing.T) {
                        defer PrintCurrentTest(t)()
@@ -898,32 +895,6 @@ func annexObjectPath(repoPath string, file string) (string, error) {
        return path.Join(repoPath, "annex", "objects", keyHashPrefix, annexKey, annexKey), nil
 }
 
-/*
-Do chmod -R +w $REPOS in order to handle https://git-annex.branchable.com/internals/lockdown/:
-
-> (The only bad consequence of this is that rm -rf .git doesn't work unless you first run chmod -R +w .git)
-
-Without, these tests can only be run once, because they reuse `gitea-repositories/`
-folder and will balk at finding pre-existing partial repos.
-*/
-func annexUnlockdown() {
-       filepath.WalkDir(setting.RepoRootPath, func(path string, d fs.DirEntry, err error) error {
-               if err == nil {
-                       // 0200 == u+w, in octal unix permission notation
-                       info, err := d.Info()
-                       if err != nil {
-                               return err
-                       }
-
-                       err = os.Chmod(path, info.Mode()|0200)
-                       if err != nil {
-                               return err
-                       }
-               }
-               return nil
-       })
-}
-
 /* like withKeyFile(), but automatically sets it the account given in ctx for use by git-annex */
 func withAnnexCtxKeyFile(t *testing.T, ctx APITestContext, callback func()) {
        os.Setenv("GIT_ANNEX_USE_GIT_SSH", "1") // withKeyFile works by setting GIT_SSH_COMMAND, but git-annex only respects that if this is set
diff --git a/modules/util/remove.go b/modules/util/remove.go
index 95024e11f..bfb9a9782 100644
--- a/modules/util/remove.go
+++ b/modules/util/remove.go
@@ -6,6 +6,7 @@ package util
 
 import (
        "os"
+       "fmt"
        "runtime"
        "syscall"
        "time"
@@ -46,6 +47,8 @@ func Remove(name string) error {
 
 // RemoveAll removes the named file or directory with at most 5 attempts.
 func RemoveAll(name string) error {
+
+       fmt.Printf("RemoveAll(%s)\n", name)
        var err error
 
        for i := 0; i < 5; i++ {
```

</details>

<details><summary>test</summary>

```
p115628@joplin:~/src/neurogitea/gitea$ make test-sqlite#TestGitAnnex.Public.Reader.SSH.Init
RemoveAll(/home/GRAMES.POLYMTL.CA/p115628/src/neurogitea/gitea/integrations/gitea-integration-sqlite/tmp/uploads)
RemoveAll(/home/GRAMES.POLYMTL.CA/p115628/src/neurogitea/gitea/integrations/gitea-integration-sqlite/data/integrations/gitea-integration-sqlite/tmp/local-repo)
=== TestGitAnnex (integrations/git_helper_for_declarative_test.go:117)
RemoveAll(/home/GRAMES.POLYMTL.CA/p115628/src/neurogitea/gitea/integrations/gitea-integration-sqlite/gitea-repositories)
=== TestGitAnnex/Public (integrations/git_annex_test.go:62)
RemoveAll(/tmp/gitea-annex-public4143071661)
RemoveAll(/tmp/index3724207135)
RemoveAll(/tmp/index2224850679)
RemoveAll(/tmp/annex-public2371856901)
RemoveAll(/tmp/key-file88687326)
=== TestGitAnnex/Public/Reader (integrations/git_annex_test.go:212)
=== TestGitAnnex/Public/Reader/SSH (integrations/git_annex_test.go:215)
RemoveAll(/tmp/key-file3426719213)
=== TestGitAnnex/Public/Reader/SSH/Init (integrations/git_annex_test.go:241)
RemoveAll(/tmp/key-file594703350)
RemoveAll(/tmp/annex-public253285415)
+++ TestGitAnnex/Public is a slow test (took 20.893163374s)
+++ TestGitAnnex is a slow test (took 20.97350748s)
PASS
```

</details>

the key lines being

> === TestGitAnnex (integrations/git_helper_for_declarative_test.go:117)
> RemoveAll(/home/GRAMES.POLYMTL.CA/p115628/src/neurogitea/gitea/integrations/gitea-integration-sqlite/gitea-repositories)